File: CodingTest/Java/Solution28.java
// https://school.programmers.co.kr/learn/courses/30/lessons/43165
package CodingTest.Java;

public class Solution28 {
    public static int answer = 0;
    public int solution(int[] numbers, int target) {
        dfs(numbers, 0, 0, target);
        return answer;
    }

    public void dfs(int[] numbers, int pointer, int result, int target) {
        if (pointer == numbers.length && result == target) {
            answer += 1;
        }
        if (!(pointer >= numbers.length)) {
            for (int i = 0; i < 2; i++) {
                if (i == 0) {
                    dfs(numbers, pointer + 1, result + numbers[pointer], target);
                } else {
                    dfs(numbers, pointer + 1, result - numbers[pointer], target);
                }
            }   
        }
    }
}

File: CodingTest/Python/solution28.py
# https://school.programmers.co.kr/learn/courses/30/lessons/118668

def solution(alp, cop, problems):
	max_a = 0
	max_c = 0
	for problem in problems:
		max_a = max(max_a, problem[0])
		max_c = max(max_c, problem[1])
	dp = [[9999 for _ in range(max_c + 1)] for _ in range(max_a + 1)]
	
	alp = min(alp, max_a)
	cop = min(cop, max_c)
	dp[alp][cop] = 0
	for i in range(alp, max_a + 1):
		for j in range(cop, max_c + 1):
			# 문제 풀이 없이 공부하는 방법
			if i < max_a:
				dp[i + 1][j] = min(dp[i + 1][j], dp[i][j] + 1)
			if j < max_c:
				dp[i][j + 1] = min(dp[i][j + 1], dp[i][j] + 1)
			# 문제 풀이로 공부하는 방법
			for problem in problems:
				if i >= problem[0] and j >= problem[1]:
					a = min(i + problem[2], max_a)
					c = min(j + problem[3], max_c)
					dp[a][c] = min(dp[a][c], dp[i][j] + problem[-1])
					
	return dp[max_a][max_c]

# BFS로 하면 효율성 박살난대